var/spack/repos/builtin/packages/py-htgettoken/package.py

```diff
@@ -0,0 +1,39 @@
+# Copyright 2013-2023 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.package import *
+
+
+class PyHtgettoken(PythonPackage):
+    """htgettoken gets OIDC authentication tokens for High Throughput Computing
+    via a Hashicorp vault server."""
+
+    homepage = "https://github.com/fermitools/htgettoken"
+
+    # htgettoken is not available on PyPi
+    url = "https://github.com/fermitools/htgettoken/archive/refs/tags/v1.16.tar.gz"
+    list_url = "https://github.com/fermitools/htgettoken/releases/"
+    git = "https://github.com/fermitools/htgettoken.git"
+
+    maintainers = ["wdconinc"]
+
+    version("1.16-33-g3788bb4", commit="3788bb4733e5e8f856cee51566df9a36cbfe097d")
+    version("1.16-20-g8b72f48", commit="8b72f4800ef99923dac99dbe0756a26266a27886")
+    # Older versions do not have a python build system
+    version(
+        "1.16",
+        sha256="984fd4746082086356d968b4db536f7b35d30e65a6589fa357caa3266cb98268",
+        deprecated=True,
+    )
+
+    depends_on("py-setuptools", type="build")
+
+    depends_on("py-gssapi", type="run")
+    depends_on("py-paramiko", type="run")
+    depends_on("py-urllib3", type="run")
+
+    def setup_run_environment(self, env):
+        if env.get("XDG_RUNTIME_DIR") and env.get("UID"):
+            env.set("BEARER_TOKEN", join_path(env.get("XDG_RUNTIME_DIR"), "bt_u" + env.get("UID")))
```

Comment:
If you're relying on these being set in the osg-ca-certs package I'm not sure if this will always be the case.

Reply:
@alalazo do you have any thoughts on this?

Reply:
Yes, you are right that this could be made more robust. Upon inspection, the package will place the token at `BEARER_TOKEN_FILE` if defined, and will read from `BEARER_TOKEN` if defined, `BEARER_TOKEN_FILE` otherwise. The default file if none of this is defined is based on `XDG_RUNTIME_DIR` or `/tmp` and `UID` or `os.getuid()`.

I think the following is more robust then (but will need `import os` at the top):
```suggestion
    def setup_run_environment(self, env):
        dir = env.get("XDG_RUNTIME_DIR") or "/tmp"
        uid = env.get("UID") or os.geteuid()
        file = join_path(dir, "bt_u" + uid)
        env.set("BEARER_TOKEN", file)
        env.set("BEARER_TOKEN_FILE", file)
```

Reply:
This still relies on env vars set by other packages, so want to get @alalazo's feedback first

Reply:
Nothing sets `XDG_RUNTIME_DIR` that is in spack, so there's nothing this package can depend on. `UID` is a bash variable, but we cannot rely on running bash either. So, maybe another way would be to ignore these locations where htgettoken itself would put the token, and just construct our own path:
```
    def setup_run_environment(self, env):
        dir = env.get("TMPDIR") or "/tmp"
        uid = os.geteuid()
        file = join_path(dir, "bt_u" + uid)
        env.set("BEARER_TOKEN", file)
        env.set("BEARER_TOKEN_FILE", file)
```

We can't really use `tempfile.mkdtemp` since this has to persist across shell sessions through the token lifetime, say 24 hours (otherwise a user has to re-authenticate for every shell session).